### ISSUE TITLE
Too demanding enum predicate check

### ISSUE THREAD

**Issue by NikVolf** [Manishearth/rust-clippy#677]
```
ethcore/src/basic_types.rs:28:1: 33:2 warning: All variants have the same prefix: `With`, #[warn(enum_variant_names)] on by default
ethcore/src/basic_types.rs:28 pub enum Seal {
ethcore/src/basic_types.rs:29 	/// The seal/signature is included.
ethcore/src/basic_types.rs:30 	With,
ethcore/src/basic_types.rs:31 	/// The seal/signature is not included.
ethcore/src/basic_types.rs:32 	Without,
ethcore/src/basic_types.rs:33 }
```

though technically correct, looks kind of naughty

**Comment by oli-obk** [comment]
on it already

**Comment by oli-obk** [comment]
ah... Well... in this situation maybe you should name the enum `WithSeal` and change the variants to `Yes` and `No`. But this is opinionated. I'm not sure how to proceed. I could create an allow-by default lint that warns in this case, and the normal lint in all other cases.

**Comment by Manishearth** [comment]
Wait, wasn't that the point of the CamelCase check?

The idea was that a prefix must be a CamelCase component; not just any prefix.

**Comment by oli-obk** [comment]
ah... good point. Yea, I only check the first variant... so if the first variant is equal to the prefix... well you get the picture

**Comment by oli-obk** [comment]
So I'll check the first two variants, so we still catch `enum Seal { With, WithOut, }`

**Comment by Manishearth** [comment]
Ah. We should also avoid cases where the CamelCase prefix is the whole enum.

I'm thinking of designing the check as something which takes the union of the first two strings (where "union" is "largest prefix which is followed by a capital character"), and then takes the union of that with the next string, and so on. And then does the same for suffixes.

The "followed by a capital character" both ensures that it's a camelcase prefix, as well as ensuring that the "prefix" isn't the whole thing.

**Comment by oli-obk** [comment]
> Ah. We should also avoid cases where the CamelCase prefix is the whole enum.

Imo something like

```rust
enum ReturnKind {
    Return(Something),
    ReturnDefault,
    ReturnNothing,
}
```

is wrong.

Also we can't do the `followed by a capital character` check, because that would break non-ascii:

```rust
enum NonCaps { //~ ERROR: All variants have the same prefix: `Prefix`
    Prefix的,
    PrefixCake,
}
```

so it would need to be "is followed by a non lowercase character". Still, if you prefer to not lint when the prefix is the same as a variant, do you mind if I add an additional allow-by-default lint for that case?

**Comment by Manishearth** [comment]
> Still, if you prefer to not lint when the prefix is the same as a variant, do you mind if I add an additional allow-by-default lint for that case?

sgtm

**Comment by Manishearth** [comment]
https://github.com/Manishearth/rust-clippy/pull/673
